submission/loglist_refresher_test.go

```diff
@@ -0,0 +1,184 @@
+		if gotEvt.Err == nil || !errRegexp.MatchString(gotEvt.Err.Error()) {
+			t.Errorf("Got error event %q did not match wanted substring %q", gotEvt.Err, errRegexp)
+		}
+	} else if gotEvt.Err != nil {
+		t.Errorf("Got error event %q while none expected", gotEvt.Err)
+	}
+}
+
+func TestNewLoglistRefresherNoFile(t *testing.T) {
+	t.Run("NoDataFile", func(t *testing.T) {
+		llr := NewLoglistRefresher("nofile.json")
+		ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+		defer cancel()
+		go llr.Run(ctx)
+		evt := <-llr.Events
+		if !strings.Contains(evt.Err.Error(), "failed to read") {
+			t.Errorf("Expected getting error event on reading non-existent file, got %q", evt.Err)
+		}
+	})
+}
+
+func TestNewLoglistRefresher(t *testing.T) {
+	testCases := []struct {
+		name      string
+		ll        string
+		wantLl    *loglist.LogList
+		errRegexp *regexp.Regexp
+	}{
+		{
+			name:   "SuccessfulRead",
+			ll:     `{"operators": [{"id":0,"name":"Google"}]}`,
+			wantLl: &loglist.LogList{Operators: []loglist.Operator{{ID: 0, Name: "Google"}}},
+		},
+		{
+			name:      "CannotParseInput",
+			ll:        `invalid`,
+			errRegexp: regexp.MustCompile("failed to parse"),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			f := createDataFile(t, "loglist.json", []byte(tc.ll))
+			defer os.Remove(f.Name())
+
+			llr := NewLoglistRefresher(f.Name())
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+			defer cancel()
+			go llr.Run(ctx)
+
+			gotEvt := <-llr.Events
+			compareEvents(t, gotEvt, tc.wantLl, tc.errRegexp)
+		})
+	}
+}
+
+func TestNewLoglistRefresherUpdate(t *testing.T) {
+	testCases := []struct {
+		name      string
+		ll        string
+		llNext    string
+		wantLl    *loglist.LogList
+		errRegexp *regexp.Regexp
+	}{
+		{
+			name:      "NoUpdate",
+			ll:        `{"operators": [{"id":0,"name":"Google"}]}`,
+			llNext:    `{"operators": [{"id":0,"name":"Google"}]}`,
+			wantLl:    nil,
+			errRegexp: nil,
+		},
+		{
+			name:      "LoglistUpdated",
+			ll:        `{"operators": [{"id":0,"name":"Google"}]}`,
+			llNext:    `{"operators": [{"id":0,"name":"GoogleOps"}]}`,
+			wantLl:    &loglist.LogList{Operators: []loglist.Operator{{ID: 0, Name: "GoogleOps"}}},
+			errRegexp: nil,
+		},
+		{
+			name:      "CannotReadInput",
+			ll:        `{"operators": [{"id":0,"name":"Google"}]}`,
+			llNext:    `invalid`,
+			errRegexp: regexp.MustCompile("failed to parse"),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			f := createDataFile(t, "loglist.json", []byte(tc.ll))
+			defer os.Remove(f.Name())
+
+			llr := NewLoglistRefresher(f.Name())
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			go llr.Run(ctx)
+			<-llr.Events
+
+			// Simulate Log list update.
+			rewriteDataFile(t, f.Name, []byte(tc.llNext))
+
+			go llr.Run(ctx)
+			// Wait for event if any.
+			time.Sleep(time.Second * 1)
```

Comment:
Rather than sleeping, which will always extend the duration of this test by 1 second, you could instead wait on a context to expire in the select below. In the case of an event being present on the channel, the test will progress quickly, otherwise they'll be a 1 second delay before it fails.

Reply:
Updated.